### PR TITLE
Qualification tool base plan parsing infrastructure

### PR DESCRIPTION
**Pull request by tgravescs**
This does not add anything to official output for the Qualification tool but it adds some classes and logic for parsing the plan and handling each exec.  This adds WholeStagecodegen, Project and Filter classes. It was split this way partially to help with multiple people being able to develop at once without colliding.

The output is just a log debug statements and added a few basic tests for current functionality.

Lots of TODOs left in code intentionally.

fixes https://github.com/NVIDIA/spark-rapids/issues/5358

### CODE REVIEW

**Comment by tgravescs** [comment]
build

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/FilterExecParser.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import com.nvidia.spark.rapids.tool.qualification.PluginTypeChecker
+
+import org.apache.spark.sql.execution.ui.SparkPlanGraphNode
+import org.apache.spark.sql.rapids.tool.AppBase
+
+case class FilterExecParser(
+    node: SparkPlanGraphNode,
+    checker: PluginTypeChecker,
+    sqlID: Long,
+    app: AppBase) extends ExecParser {
```
Comment (nartal1) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/FilterExecParser.scala:28]: Nit: I don't see `app` being used here and in `ProjectExecParser` . Can we remove it? If it is kept to be used later or to keep it in sync with `WholeStageExecParser` , then I am fine with it.

File: tools/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala [hunk 2]
```diff
@@ -217,6 +218,30 @@ class QualificationAppInfo(
       val (allComplexTypes, nestedComplexTypes) = reportComplexTypes
       val problems = getAllPotentialProblems(getPotentialProblemsForDf, nestedComplexTypes)
 
+      val planInfos = sqlPlans.map { case (id, plan) =>
```
Comment (nartal1) [inline tools/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala:221]: The purpose of this variable for now is to log the total duration. Is it okay to assume `planInfos` would be used in qualification output later?

File: tools/src/test/scala/com/nvidia/spark/rapids/tool/planparser/SqlPlanParserSuite.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          val projects = allChildren.filter(_.exec == "ProjectExec")
+          assert(projects.forall(_.speedupFactor == 2))
+          assert(projects.forall(_.isSupported == true))
+          assert(projects.forall(_.children.isEmpty))
+          assert(projects.forall(_.duration.isEmpty))
+        }
+      }
+    }
+  }
+}
```
Comment (nartal1) [inline tools/src/test/scala/com/nvidia/spark/rapids/tool/planparser/SqlPlanParserSuite.scala:90]: Nit: newline

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/FilterExecParser.scala (hunk 1 above)
Comment (tgravescs) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/FilterExecParser.scala:28]: we can remove it for now if we need it later for expressions we can add it back. thanks for catching

**Review by nartal1 — APPROVED** [review]
LGTM.  Thanks @tgravescs. This PR will help others to add  different Execs and Expr for speedUpFactor.

**Review by amahussein — APPROVED** [review]
LGTME

**Comment by tgravescs** [comment]
build

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        ProjectExecParser(p, checker, sqlID).parse
+      case o =>
+        logDebug(s"other graph node ${node.name} desc: ${node.desc} id: ${node.id}")
+        ArrayBuffer(ExecInfo(sqlID, o.name, expr = "", 1, duration = None, o.id,
+          isSupported = false, None))
+    }
+  }
+
+  // integer division if we want to roundup need to change
+  def average(arr: ArrayBuffer[Int]): Int = if (arr.isEmpty) 1 else arr.sum/arr.size
```
Comment (gerashegalov) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:97]: if the sum of elements is 0 for an empty set, would avg(empty)=0 make more sense?

nit: spaces around `/`

Comment (tgravescs) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:97]: In  this case it's used with a speedup factor and applied to a duration , so your duration isn't going to go to 0, its going to stay the same so for empty array 1 makes sense, if its non-empty the sum should never be 0 as the speed-factor for individual things would be at least 1.

 Probably need a better name and documentation here so its clear and not used for other things

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
Note parsed plan logging looks something like:

```
22/04/28 16:43:37 WARN QualificationAppInfo: sqlID: 24, exec: exec: Execute InsertIntoHadoopFsRelationCommand, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 0, isSupported: false, children:
exec: TakeOrderedAndProject, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 1, isSupported: false, children:
exec: WholeStageCodegen (14), expr: WholeStageCodegen (14), sqlID: 24 , speedupFactor: 1, duration: Some(301876), nodeId: 2, isSupported: true, children:
       exec: Project, expr: , sqlID: 24 , speedupFactor: 2, duration: None, nodeId: 3, isSupported: true, children:
       exec: SortMergeJoin, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 4, isSupported: false, children:
exec: WholeStageCodegen (11), expr: WholeStageCodegen (11), sqlID: 24 , speedupFactor: 1, duration: Some(398983), nodeId: 5, isSupported: false, children:
       exec: Sort, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 6, isSupported: false, children:
exec: Exchange, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 7, isSupported: false, children:
exec: WholeStageCodegen (10), expr: WholeStageCodegen (10), sqlID: 24 , speedupFactor: 1, duration: Some(509116), nodeId: 8, isSupported: true, children:
       exec: Project, expr: , sqlID: 24 , speedupFactor: 2, duration: None, nodeId: 9, isSupported: true, children:
       exec: BroadcastHashJoin, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 10, isSupported: false, children:
       exec: Project, expr: , sqlID: 24 , speedupFactor: 2, duration: None, nodeId: 11, isSupported: true, children:
       exec: SortMergeJoin, expr: , sqlID: 24 , speedupFactor: 1, duration: None, nodeId: 12, isSupported: false, children:
```

**Review by gerashegalov — APPROVED** [review]
LGTM

**Review by nartal1 — APPROVED** [review]

**Pull request merged**